src/wayland/dmabuf/dispatch.rs

```diff
@@ -0,0 +1,250 @@
+                        format!("Plane index {} is already set.", plane_idx),
+                    );
+                    return;
+                }
+
+                let modifier = ((modifier_hi as u64) << 32) + (modifier_lo as u64);
+                planes.push(Plane {
+                    fd: Some(fd),
+                    plane_idx,
+                    offset,
+                    stride,
+                    modifier: Modifier::from(modifier),
+                });
+            }
+
+            zwp_linux_buffer_params_v1::Request::Create {
+                width,
+                height,
+                format,
+                flags,
+            } => {
+                if let Some(dmabuf) = data.create_dmabuf(dh, params, width, height, format, flags) {
+                    if state.dmabuf_state().globals.get(&data.id).is_some() {
+                        if !state.dmabuf_imported(&DmabufGlobal { id: data.id }, dmabuf.clone()) {
+                            // Failed to import
+                            params.failed(dh);
+                            return;
+                        }
+
+                        match Buffer::create_buffer::<D, _>(dh, client, dmabuf) {
+                            Ok((wl_buffer, _)) => {
+                                params.created(dh, &wl_buffer);
+                            }
+
+                            Err(_) => {
+                                slog::error!(
+                                    data.logger,
+                                    "failed to create protocol object for \"create\" request"
+                                );
+                                // Failed to import since the buffer protocol object could not be created.
+                                params.failed(dh);
+                            }
+                        }
+                    } else {
+                        // If the dmabuf global was destroyed, we cannot import any buffers.
+                        params.failed(dh);
+                    }
+                }
+            }
+
+            zwp_linux_buffer_params_v1::Request::CreateImmed {
+                buffer_id,
+                width,
+                height,
+                format,
+                flags,
+            } => {
+                // Client is killed if the if statement is not taken.
+                if let Some(dmabuf) = data.create_dmabuf(dh, params, width, height, format, flags) {
+                    if state.dmabuf_state().globals.get(&data.id).is_some() {
+                        if !state.dmabuf_imported(&DmabufGlobal { id: data.id }, dmabuf.clone()) {
+                            // Buffer import failed.
+                            //
+                            // The reason for failure must be a renderer/implementation specific error.
+                            // This means we create an invalid wl_buffer and tell the client import failed.
+                            params.failed(dh);
```

Comment:
The protocol allows us to either create an invalid buffer (as is done here) or kill the client with a protocol error. Could you explain why you chose the former rather than the latter?

I'm not saying we should do the latter, I just would like to know your reasoning for choosing this.

Reply:
My rationale is as follows:
- Invalid parameters should result in a protocol error (fulfilling option one in the protocol)
- implementation specific errors should just result in import failure assuming all parameters are valid. 

I decided to read into this further, there are really two errors the renderer could have during import:
1. The number of planes is not valid for the format. This is something the renderer has to know and cannot be communicated easily to the frontend without some changes. Vulkan for example can get this info from https://www.khronos.org/registry/vulkan/specs/1.3-extensions/man/html/VkDrmFormatModifierPropertiesEXT.html
2. Implementation failure, the device was hotplugged mid import.

Since there are two cases, import should probably return a result so that an invalid plane count for a format can result in the client being killed.

Reply:
Per matrix discussion, the protocol heavily hints that create_immed should just cause client failure.